Firestore/core/src/firebase/firestore/model/field_path.cc

```diff
@@ -0,0 +1,165 @@
+  const bool needs_escaping = !IsValidIdentifier(escaped);
+  if (needs_escaping) {
+    escaped.insert(escaped.begin(), '`');
+    escaped.push_back('`');
+  }
+  return escaped;
+}
+
+}  // namespace
+
+FieldPath FieldPath::ParseServerFormat(const absl::string_view path) {
+  // TODO(b/37244157): Once we move to v1beta1, we should make this more
+  // strict. Right now, it allows non-identifier path components, even if they
+  // aren't escaped. Technically, this will mangle paths with backticks in
+  // them used in v1alpha1, but that's fine.
+
+  SegmentsT segments;
+  std::string segment;
+  segment.reserve(path.size());
+
+  // string_view doesn't have a c_str() method, because it might not be
+  // null-terminated. Assertions expect C strings, so construct std::string on
+  // the fly, so that c_str() might be called on it.
+  const auto to_string = [](const absl::string_view view) {
+    return std::string{view.data(), view.data() + view.size()};
+  };
+  const auto finish_segment = [&segments, &segment, &path, &to_string] {
+    FIREBASE_ASSERT_MESSAGE(
+        !segment.empty(),
+        "Invalid field path (%s). Paths must not be empty, begin with "
+        "'.', end with '.', or contain '..'",
+        to_string(path).c_str());
+    // Move operation will clear segment, but capacity will remain the same
+    // (not, strictly speaking, required by the standard, but true in practice).
+    segments.push_back(std::move(segment));
+  };
+
+  // Inside backticks, dots are treated literally.
+  bool inside_backticks = false;
+  // Whether to treat '\' literally or as an escape character.
+  bool escaped_character = false;
+  for (const char c : path) {
+    // std::string (and string_view) may contain embedded nulls. For full
+    // compatibility with Objective C behavior, finish upon encountering the
+    // first terminating null.
+    if (c == '\0') {
+      break;
+    }
+    if (escaped_character) {
+      escaped_character = false;
+      segment += c;
+      continue;
+    }
+
+    switch (c) {
+      case '.':
+        if (!inside_backticks) {
+          finish_segment();
+        } else {
+          segment += c;
+        }
+        break;
+
+      case '`':
+        inside_backticks = !inside_backticks;
+        break;
+
+      case '\\':
+        escaped_character = true;
+        break;
+
+      default:
+        segment += c;
+        break;
+    }
+  }
+  finish_segment();
+
+  FIREBASE_ASSERT_MESSAGE(!inside_backticks, "Unterminated ` in path %s",
+                          to_string(path).c_str());
+  // TODO(b/37244157): Make this a user-facing exception once we
+  // finalize field escaping.
+  FIREBASE_ASSERT_MESSAGE(!escaped_character,
+                          "Trailing escape characters not allowed in %s",
+                          to_string(path).c_str());
+
+  return FieldPath{std::move(segments)};
+}
+
+FieldPath FieldPath::KeyFieldPath() {
```

Comment:
I don't think we need a shared instance of the key field path in C++, and I don't think we need EmptyPath at all, because the default constructor does the same thing.

Reply:
Done.